## Source acquisition and file classification

- This layer code should be framework-agnostic and should be run it before Docling/LangChain.
- Docling is strong for downstream parsing because it supports many formats and preserves text, tables, pictures, hierarchy, layout boxes, and provenance in a unified document model.
- OCRmyPDF is a good OCR fallback because it adds searchable text layers to scanned PDFs and is tolerant of born-digital PDFs that do not need OCR.
- LangChain can then consume the cleaned outputs later through loaders such as DoclingLoader, Unstructured, JSONLoader, BSHTMLLoader, and CSVLoader.

## PDF layout and reading-order extraction



**Can docling handle multi col text in correct order?**

In [31]:
from pprint import pprint
from pathlib import Path
from docling.document_converter import DocumentConverter

In [4]:
data_folder_path = Path("/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named")
file_path = data_folder_path / "2021-winter-spring-mono-lake-newsletter-medium-mix-multi-col-text.pdf"
file_path

PosixPath('/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named/2021-winter-spring-mono-lake-newsletter-medium-mix-multi-col-text.pdf')

In [5]:
converter = DocumentConverter()
doc = converter.convert(file_path).document

[INFO] 2026-04-11 13:44:14,766 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-11 13:44:15,022 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-11 13:44:15,023 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-11 13:44:15,196 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-11 13:44:15,255 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer.onnx
[INFO] 2026-04-11 13:44:15,256 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer

In [6]:
doc.export_to_markdown()

"<!-- image -->\n\n## N E W S L E T T E R\n\nV h i s d p  %  R o p h i f   1 / 1 0\n\nRspdWl PdrsnpWshnm sn :cuWmbd        Rntsg SteW Pdnodmr        CdBgWladWt Onmcr Opnidbs        Dhdkc RdlhmWpr\n\nBNUDQ OGNSN BNTQSDRW NE C9UHC I- FTADQMHBJ&amp; P.CMANTROCPCS BNI\n\nMNQ9 KHUHMFRSNM\n\ns I write this note it feels like spring is on its way. The sun's a little warmer and the birds a little more vocal. It feels like we're turning a corner in this pandemic too. And there's good news about Mono Basin stream restoration.\n\nAll we have to do is wait for spring. But unlike the changing season, almost everything else-certainly everything in this issue of the Newsletter -happens because of human eff  orts. Sometimes those eff  orts are monumental; more often they're incremental and together, they add up.\n\nThat's certainly the case with stream restoration. We have made incremental eff  orts for years against DWP's deliberate delay tactics. Persistent phone calls, letters, and meetings are fi 

In [9]:
type(doc)

docling_core.types.doc.document.DoclingDocument

In [12]:
docdict = dict(doc)
docdict

{'schema_name': 'DoclingDocument',
 'version': '1.10.0',
 'name': '2021-winter-spring-mono-lake-newsletter-medium-mix-multi-col-text',
 'origin': DocumentOrigin(mimetype='application/pdf', binary_hash=7132794411961773591, filename='2021-winter-spring-mono-lake-newsletter-medium-mix-multi-col-text.pdf', uri=None),
 'furniture': GroupItem(self_ref='#/furniture', parent=None, children=[], content_layer=<ContentLayer.FURNITURE: 'furniture'>, meta=None, name='_root_', label=<GroupLabel.UNSPECIFIED: 'unspecified'>),
 'body': GroupItem(self_ref='#/body', parent=None, children=[RefItem(cref='#/pictures/0'), RefItem(cref='#/texts/2'), RefItem(cref='#/texts/3'), RefItem(cref='#/texts/4'), RefItem(cref='#/texts/5'), RefItem(cref='#/texts/6'), RefItem(cref='#/texts/7'), RefItem(cref='#/texts/8'), RefItem(cref='#/texts/9'), RefItem(cref='#/texts/10'), RefItem(cref='#/texts/11'), RefItem(cref='#/texts/12'), RefItem(cref='#/texts/13'), RefItem(cref='#/texts/14'), RefItem(cref='#/pictures/1'), RefItem

In [13]:
docdict.keys()

dict_keys(['schema_name', 'version', 'name', 'origin', 'furniture', 'body', 'groups', 'texts', 'pictures', 'tables', 'key_value_items', 'form_items', 'field_regions', 'field_items', 'pages'])

In [16]:
type(docdict['pages']), docdict['pages']

(dict,
 {1: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=1),
  2: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=2),
  3: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=3),
  4: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=4),
  5: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=5),
  6: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=6),
  7: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=7),
  8: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=8),
  9: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=9),
  10: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=10),
  11: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=11),
  12: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=12),
  13: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=13

In [17]:
docdict['pages'][3]

PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=3)

**How can I print content of a specific page?**
- 'pages' contains a dict which is a list of PageItem objects, there is no page content, it also says image=None
- maybe 'texts' will contain those info

In [21]:
docdict['texts']

[TextItem(self_ref='#/texts/0', parent=RefItem(cref='#/pictures/0'), children=[], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.TEXT: 'text'>, prov=[ProvenanceItem(page_no=1, bbox=BoundingBox(l=85.50278399999999, t=695.9567095955999, r=265.999018848, b=629.9933517955529, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 4))], source=[], comments=[], orig='LNMN', text='LNMN', formatting=None, hyperlink=None),
 TextItem(self_ref='#/texts/1', parent=RefItem(cref='#/pictures/0'), children=[], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.TEXT: 'text'>, prov=[ProvenanceItem(page_no=1, bbox=BoundingBox(l=340.666992, t=695.9567095955999, r=524.890242576, b=629.9933517955529, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 4))], source=[], comments=[], orig='K IA', text='K IA', formatting=None, hyperlink=None),
 SectionHeaderItem(self_ref='#/texts/2', parent=RefItem(cref='#/body'), children=[], content_

In [25]:
len(docdict['texts']), type(docdict['texts'])

(695, list)

- There is a list of 695 TextItem s
- Let's observe the item at idx 0

In [27]:
dict(docdict['texts'][0])

{'self_ref': '#/texts/0',
 'parent': RefItem(cref='#/pictures/0'),
 'children': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'meta': None,
 'label': <DocItemLabel.TEXT: 'text'>,
 'prov': [ProvenanceItem(page_no=1, bbox=BoundingBox(l=85.50278399999999, t=695.9567095955999, r=265.999018848, b=629.9933517955529, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 4))],
 'source': [],
 'comments': [],
 'orig': 'LNMN',
 'text': 'LNMN',
 'formatting': None,
 'hyperlink': None}

- Now, let's observe all the text items from page no 3

In [32]:
for textitem in docdict['texts']:
    textitem = dict(textitem)
    if textitem['prov'][0].page_no == 3:
        print(f"{textitem['self_ref']} -----------")
        pprint(textitem)
        print()

#/texts/38 -----------
{'children': [],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'formatting': None,
 'hyperlink': None,
 'label': <DocItemLabel.SECTION_HEADER: 'section_header'>,
 'level': 1,
 'meta': None,
 'orig': 'RspdWl pdrsnpWshnm onhrdc sn kdWo enpvWpc hm 1-10',
 'parent': RefItem(cref='#/body'),
 'prov': [ProvenanceItem(page_no=3, bbox=BoundingBox(l=133.858992, t=742.231919202, r=478.142775552, b=682.862469940578, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 49))],
 'self_ref': '#/texts/38',
 'source': [],
 'text': 'RspdWl pdrsnpWshnm onhrdc sn kdWo enpvWpc hm 1-10'}

#/texts/39 -----------
{'children': [],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'formatting': None,
 'hyperlink': None,
 'label': <DocItemLabel.TEXT: 'text'>,
 'meta': None,
 'orig': 'Nudpbnlhmf CSO cdkZx bkdZpr oZsg sn Zbshnm ax Edneepdx LbPthkjhm',
 'parent': RefItem(cref='#/body'),
 'prov': [ProvenanceItem(page_no=3, bbox=BoundingBox(l=194.0

**Observations**
- Multicolumn and multiblocks of texts
- Among all text item blocks 'DocItemLabel.TEXT' are extracted correctly
- But 'DocItemLabel.CAPTION', 'DocItemLabel.PAGE_FOOTER' or 'DocItemLabel.SECTION_HEADER' are not extracted correctly

Let's check some other document: